### PR TITLE
[SW-4] Implement/Integrate properties parser

### PR DESCRIPTION
**Pull request by AndriyBosik**

### CODE REVIEW

**Review by NikolayProkopchuk — APPROVED** [review]

File: src/main/java/com/example/property/service/impl/ConfigurationPropertiesServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    @Override
+    public ConfigurationProperties getProperties() {
+        return getProperties(DEFAULT_LOCATION);
+    }
+
+    @Override
+    public ConfigurationProperties getProperties(String location) {
+        URL resource = getClass().getClassLoader().getResource(location);
+        if (resource == null) {
+            throw new InvalidConfigurationPropertiesLocation("Properties by location " + location + " do not exists");
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/ConfigurationPropertiesServiceImpl.java:44]: let's move into Constants

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        int searchFrom = 0;
+        int envPatternStartFrom;
+        int plainValueStartFrom = 0;
+        StringBuilder value = new StringBuilder();
+        while ((envPatternStartFrom = propertyValue.indexOf(ENVIRONMENT_START_SUBSTRING, searchFrom)) > -1) {
+            int envValueStartFrom = envPatternStartFrom + ENVIRONMENT_START_SUBSTRING.length();
+            int closingCurlyBraceIndex = findBalancedBrace(propertyValue, envValueStartFrom);
+            if (closingCurlyBraceIndex < 0) {
+                throw new InvalidEnvironmentVariable("Invalid property value: '" + propertyValue + "'. Brackets are not balanced");
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:23]: let's move into Constants

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            return null;
+        }
+        int searchFrom = 0;
+        int envPatternStartFrom;
+        int plainValueStartFrom = 0;
+        StringBuilder value = new StringBuilder();
+        while ((envPatternStartFrom = propertyValue.indexOf(ENVIRONMENT_START_SUBSTRING, searchFrom)) > -1) {
+            int envValueStartFrom = envPatternStartFrom + ENVIRONMENT_START_SUBSTRING.length();
+            int closingCurlyBraceIndex = findBalancedBrace(propertyValue, envValueStartFrom);
+            if (closingCurlyBraceIndex < 0) {
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:22]: let's avoid magic numbers

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    @Override
+    public String parseStringValue(String propertyValue) {
+        if (StringUtils.isBlank(propertyValue)) {
+            return null;
+        }
+        int searchFrom = 0;
+        int envPatternStartFrom;
+        int plainValueStartFrom = 0;
+        StringBuilder value = new StringBuilder();
+        while ((envPatternStartFrom = propertyValue.indexOf(ENVIRONMENT_START_SUBSTRING, searchFrom)) > -1) {
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:19]: let's avoid magic numbers

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            int envValueStartFrom = envPatternStartFrom + ENVIRONMENT_START_SUBSTRING.length();
+            int closingCurlyBraceIndex = findBalancedBrace(propertyValue, envValueStartFrom);
+            if (closingCurlyBraceIndex < 0) {
+                throw new InvalidEnvironmentVariable("Invalid property value: '" + propertyValue + "'. Brackets are not balanced");
+            }
+            searchFrom = closingCurlyBraceIndex;
+            if (envPatternStartFrom - plainValueStartFrom > 0) {
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:26]: same

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                value.append(propertyValue, plainValueStartFrom, envPatternStartFrom);
+            }
+            value.append(parseEnvironmentValue(propertyValue.substring(envValueStartFrom, closingCurlyBraceIndex)));
+            plainValueStartFrom = closingCurlyBraceIndex + 1;
+        }
+        if (propertyValue.length() - 1 - plainValueStartFrom > 0) {
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:32]: same

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    private int findBalancedBrace(String propertyValue, int envStart) {
+        int notBalancesBraces = 1;
+        for (int i = envStart; i < propertyValue.length(); i++) {
+            char symbol = propertyValue.charAt(i);
+            if (symbol == '{') {
+                notBalancesBraces++;
+            } else if (symbol == '}') {
+                notBalancesBraces--;
+            }
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:52]: what do you think about switch-case block?

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private String parseEnvironmentValue(String rawValue) {
+        rawValue = StringUtils.strip(rawValue);
+        if (rawValue == null) {
+            throw new InvalidEnvironmentVariable("Null property value");
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:63]: string constant

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        int semicolonIndex = rawValue.indexOf(":");
+        if (semicolonIndex == 0) {
+            throw new InvalidEnvironmentVariable("Invalid value for environment variable: " + rawValue);
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:67]: string constant

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        return -1;
+    }
+
+    private String parseEnvironmentValue(String rawValue) {
+        rawValue = StringUtils.strip(rawValue);
+        if (rawValue == null) {
+            throw new InvalidEnvironmentVariable("Null property value");
+        }
+        int semicolonIndex = rawValue.indexOf(":");
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:65]: put into constant variable

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        if (semicolonIndex == 0) {
+            throw new InvalidEnvironmentVariable("Invalid value for environment variable: " + rawValue);
+        }
+        boolean semicolonPresent = semicolonIndex >= 1;
+        String envName = rawValue.substring(0, semicolonPresent ? semicolonIndex : rawValue.length());
+        String defaultValue = semicolonPresent ? rawValue.substring(semicolonIndex + 1) : null;
+        String systemEnvValue = StringUtils.strip(System.getenv(envName));
+        boolean blankSystemEnvValue = StringUtils.isBlank(systemEnvValue);
+        if (blankSystemEnvValue && defaultValue == null) {
+            throw new InvalidEnvironmentVariable("Environment value does not exist and no default value had been set");
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:75]: string constant

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    private String parseEnvironmentValue(String rawValue) {
+        rawValue = StringUtils.strip(rawValue);
+        if (rawValue == null) {
+            throw new InvalidEnvironmentVariable("Null property value");
+        }
+        int semicolonIndex = rawValue.indexOf(":");
+        if (semicolonIndex == 0) {
+            throw new InvalidEnvironmentVariable("Invalid value for environment variable: " + rawValue);
+        }
+        boolean semicolonPresent = semicolonIndex >= 1;
```
Comment (ivlytvynets) [inline src/main/java/com/example/property/service/impl/PropertyParserImpl.java:69]: let's avoid magic numbers

File: src/main/java/com/example/property/service/impl/ConfigurationPropertiesServiceImpl.java (hunk 1 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/ConfigurationPropertiesServiceImpl.java:44]: fixed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 2 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:23]: fixed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 3 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:22]: discussed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 4 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:19]: discussed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 5 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:26]: refactored to `enenvPatternStartFrom > plainValueStartFromv`

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 6 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:32]: extracted to a separated variable as discussed on a call

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 7 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:52]: replaced with switch-case

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 8 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:63]: fixed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 9 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:67]: fixed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 10 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:65]: fixed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 11 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:75]: fixed

File: src/main/java/com/example/property/service/impl/PropertyParserImpl.java (hunk 12 above)
Comment (AndriyBosik) [reply src/main/java/com/example/property/service/impl/PropertyParserImpl.java:69]: discussed

**Review by ivlytvynets — APPROVED** [review]

**Pull request merged**